Update CONTRIBUTING.md for crate creation. (#29637)
solana-labs/solana: CONTRIBUTING.md

Instead, add a comment that mentions the usernames that you would like a review
from. Ask explicitly what you would like feedback on.

## Crate Creation

> If your PR includes a new crate, it is required to publish its v0.0.1 version
> before the PR can be merged.  Here're the steps:

Reviewer: ```suggestion
If your PR includes a new crate, you must publish its v0.0.1 version
before the PR can be merged.  Here are the steps:
```

* Create a sub-directory for your new crate.
* Under the newly-created directory, create a Cargo.toml file.  Below is an
  example template:

```
[package]
name = "solana-<PACKAGE_NAME>"
version = "0.0.1"
description = "<DESCRIPTION>"
authors = ["Solana Maintainers <maintainers@solana.foundation>"]
repository = "https://github.com/solana-labs/solana"
homepage = "https://solana.com/"
documentation = "https://docs.rs/solana-<PACKAGE_NAME>"
license = "Apache-2.0"
edition = "2021"
```

> * Submit the PR for initial review.  At this moment you should see the crate
>   check fails because the newly created crate has not yet known in public.

Reviewer: ```suggestion
* Submit the PR for initial review.  You should see the crate-check CI job
  fails because the newly created crate is not yet published.
```

> * Once all review feedback has been addressed.  Publish v0.0.1 of the crate
>   under your account and transfer the crate ownership to solana-grimes.

Reviewer: ```suggestion
* Once all review feedback has been addressed, publish v0.0.1 of the crate
  under your personal crates.io account, and then transfer the crate ownership to solana-grimes.
```

  https://crates.io/policies#package-ownership

> * After publication went successful, update the PR by replacing v0.0.1 version
>   with the correct version.  At this time you should see the crate check passes,
>   and your published crate should be available under https://crates.io/crates/.

Reviewer: ```suggestion
* After successful publication, update the PR by replacing the v0.0.1 version number
  with the correct version.  At this time you should see the crate-check CI job passes,
  and your published crate should be available under https://crates.io/crates/.
```

Instead, add a comment that mentions the usernames that you would like a review
from. Ask explicitly what you would like feedback on.

## Crate Creation

If your PR includes a new crate, it is required to publish its v0.0.1 version
before the PR can be merged.  Here're the steps:

* Create a sub-directory for your new crate.
* Under the newly-created directory, create a Cargo.toml file.  Below is an
  example template:

```
[package]
name = "solana-<PACKAGE_NAME>"
version = "0.0.1"
description = "<DESCRIPTION>"
authors = ["Solana Maintainers <maintainers@solana.foundation>"]

> repository = "https://github.com/solana-labs/<REPO>"

Reviewer: ```suggestion
repository = "https://github.com/solana-labs/solana"
```
Since this doc is in the `solana` monorepo...

homepage = "https://solana.com/"
documentation = "https://docs.rs/solana-<PACKAGE_NAME>"
license = "Apache-2.0"
edition = "2021"
```

* Submit the PR for initial review.  At this moment you should see the crate
  check fails because the newly created crate has not yet known in public.

* Once all review feedback has been addressed.  The next step is to publish
  v0.0.01 of your crate.  This can be done by one of the following options:

   * Find a Solana engineer to publish the crate under Solana's crate account:
     https://crates.io/users/solana-grimes.

   * (Preferred) Publish the crate under your account, and transfer the crate

>      ownership to solana-grimes.  https://crates.io/policies#package-ownership

Reviewer: I lean toward just the ownership-transfer approach

Instead, add a comment that mentions the usernames that you would like a review
from. Ask explicitly what you would like feedback on.

## Crate Creation

If your PR includes a new crate, it is required to publish its v0.0.1 version
before the PR can be merged.  Here're the steps:

* Create a sub-directory for your new crate.
* Under the newly-created directory, create a Cargo.toml file.  Below is an
  example template:

```
[package]
name = "solana-<PACKAGE_NAME>"
version = "0.0.1"
description = "<DESCRIPTION>"
authors = ["Solana Maintainers <maintainers@solana.foundation>"]
repository = "https://github.com/solana-labs/<REPO>"
homepage = "https://solana.com/"
documentation = "https://docs.rs/solana-<PACKAGE_NAME>"
license = "Apache-2.0"
edition = "2021"
```

* Submit the PR for initial review.  At this moment you should see the crate
  check fails because the newly created crate has not yet known in public.

* Once all review feedback has been addressed.  We will send you the token

>   for publishing crate for Solana.  Run the following command under your crate

Reviewer 2: do we share the token to outside contributors 🤔 I think maybe we can publish the crate for them

Reviewer 3: hmm, yeah maybe a better way is to first publish the crate with your own crates.io user and then transfer the ownership to the https://crates.io/users/solana-grimes user via a discussion on #releng

Author: Yep, we definitely shouldn't share the token with external contributors as it should be treated as our "private key".

I will include the above two options (publish the crate for them, or ownership transfer).

Reviewer 3: I kinda think the external contributor path is better for "internal" contributors as well actually. Sharing that token around isn't exactly a best practice. 

Longer term we want to remove as much of the separation as possible between internal and  external contributor anyway

Author: Make sense.  Btw, when the crate is first published under a personal account, does the crate check verify whether it is under Solana or not?  Just want to make sure we don't accidentally merge some PR whose crate isn't published under Solana.

Reviewer 3: yep, owner is checked: https://github.com/solana-labs/solana/blob/3d63f93bb0346886d0062576f3382bf0e0bc9572/ci/check-crates.sh#L24

  directory to publish v0.0.1 of your new crate.  Note that to publish v0.0.1,

>   all its dependencies must use a published version.  Otherwise the publication

Reviewer: Typically, the v0.0.1 crate should be empty, with no dependencies at all. Should we specify that?
